examples/step-74/step-laplace-FEIV.cc

```diff
@@ -0,0 +1,1245 @@
+                                                   difference_per_cell,
+                                                   VectorTools::L2_norm);
+      std::cout << "   Error in the L2 norm       :     " << L2_error << std::endl;
+      
+      
+
+    }
+  }
+  template <int dim>
+  void AdvectionProblem<dim>::compute_error_estimator()
+  {
+    typedef decltype(dof_handler.begin_active()) Iterator;
+    const RHS<dim> rhs_function(test_case);
+    const Viscosity<dim> viscosity_function;
+    const Solution<dim> boundary_function(test_case);
+    estimated_error_per_cell.reinit(triangulation.n_active_cells());
+
+    auto cell_worker = [&] (const Iterator &cell, 
+                            ScratchData1<dim> &scratch_data, 
+                            CopyData &copy_data)
+    {
+      FEValues<dim> &fe_v = scratch_data.fe_values;
+      fe_v.reinit(cell);
+      
+      copy_data.cell_index = cell->active_cell_index();
+      
+      const auto & q_points = fe_v.get_quadrature_points();
+      const unsigned int n_q_points = q_points.size();
+      const std::vector<double> &JxW = fe_v.get_JxW_values ();
+
+      std::vector<Tensor<2, dim>> hessians(n_q_points);
+      fe_v.get_function_hessians(solution, hessians);
+
+      std::vector<double> nu (n_q_points);
+      viscosity_function.value_list (fe_v.get_quadrature_points(), nu);
+      std::vector<double> rhs (n_q_points);
+      rhs_function.value_list (q_points, rhs);
+
+      double residual_norm_square = 0;
+      for (unsigned int point=0; point<n_q_points; ++point)
+        {
+          const double residual = rhs[point] + nu[point]*trace(hessians[point]);
+          residual_norm_square += residual*residual*JxW[point];
+        }
+        copy_data.value =cell->diameter()*std::sqrt(  residual_norm_square);
+    };
+
+    auto boundary_worker = [&] (const Iterator &cell, 
+                                const unsigned int &face_no, 
+                                ScratchData1<dim> &scratch_data, 
+                                CopyData &copy_data)
+    {
+      scratch_data.fe_interface_values.reinit(cell, face_no);
+      const FEFaceValuesBase<dim> &fe_fv =
+      scratch_data.fe_interface_values.get_fe_face_values(0);
+
+      const auto &q_points = fe_fv.get_quadrature_points();
+      const unsigned n_q_points = q_points.size();
+      
+      const std::vector<double> &JxW = fe_fv.get_JxW_values ();
+      
+      std::vector<double> nu (n_q_points);
+      viscosity_function.value_list (q_points, nu);
+
+      std::vector<double> g(n_q_points);
+      boundary_function.value_list (q_points, g);
+
+      std::vector<double> sol_u(n_q_points);
+      fe_fv.get_function_values(solution, sol_u);
+
+      const double degree = std::max(1.0, static_cast<double> (fe_fv.get_fe().degree));
+      const double extent1 = cell->extent_in_direction(GeometryInfo<dim>::unit_normal_direction[face_no]);
+      const double penalty = 4.0 * degree * (degree+1.0) / (extent1);
```

Comment:
this penalty parameter design should move into a function instead of copying it here.